### PR TITLE
Api

### PR DESCRIPTION
**Pull request by azgas**
Api project with http get method GetAvailableHotelsIds and http post MakeReservation

### CODE REVIEW

**Comment by jacekmlynek** [comment]
Nice minimalistic intgeration

File: HotelApi/Startup.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+namespace HotelApi
+{
+    public class Startup
+    {
+        public void Configuration(IAppBuilder app)
+        {
+            HttpConfiguration config = new HttpConfiguration();
+            config.Routes.MapHttpRoute(
+                name: "DefaultApi",
+                routeTemplate: "api/{controller}/{action}",
```
Comment (jacekmlynek) [inline HotelApi/Startup.cs]: Couple of things here:
* usually in Apis we do not using action name we rather follow convention (controller method start with HTTP verb name)described [here](https://docs.microsoft.com/en-us/aspnet/web-api/overview/web-api-routing-and-actions/routing-in-aspnet-web-api)
* if you have something like info page it would be nice to add default controller and maybe action
* you do not need Id param, you have propably copied it from some external example - could you please get rid of this param.

File: HotelApi/Controllers/ReservationController.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        public ReservationController()
+        {
+            IHotelFactory factory = new HotelFactory();
+            IPaymentService paymentService = new PaymentService();
+            IBookingService bookingService = new BookingService();
+            ILogger logger = new ConsoleLogger();
+            _manager = new HotelManager(factory, paymentService, bookingService, logger);
+        }
+
+        [HttpGet]
```
Comment (jacekmlynek) [inline HotelApi/Controllers/ReservationController.cs]: Please note:

If you will follow convention you do not need to use HttpGet attribute

File: HotelApi/Controllers/ReservationController.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        public JsonResult<List<int>> GetAvailableHotelsIds()
+        {
+            return Json(_manager.PresentAvailableHotels());
+        }
+
+        [HttpPost]
+        public JsonResult<ReservationResult> MakeReservation(int hotelId, double price, int creditCardNumber,
```
Comment (jacekmlynek) [inline HotelApi/Controllers/ReservationController.cs]: Should consider:

If it is a json api or one of the supported format is json why not use it to deserialization as well. It is mean that instead of multiple attributes you can create object e.g. ReservationRequest, which  may be simple DTO with DataContract attributes to handle serialization properly. It is a lot of stuff in google [just one of the first](https://stackoverflow.com/questions/22492821/webapi-deserializing-and-serializing-alternate-property-names)

File: HotelApi/Program.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+using Microsoft.Owin.Hosting;
+
+namespace HotelApi
+{
+    class Program
+    {
+        const string Url = "http://localhost:9000";
```
Comment (jacekmlynek) [inline HotelApi/Program.cs]: Should have:

Why you have defined const outside the method if it will be used just inside one method?

File: HotelApi/Controllers/ReservationController.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace HotelApi.Controllers
+{
+    public class ReservationController : ApiController
+    {
+        private readonly HotelManager _manager;
+
+        public ReservationController()
+        {
+            IHotelFactory factory = new HotelFactory();
```
Comment (jacekmlynek) [inline HotelApi/Controllers/ReservationController.cs]: Nice to have:

Could you please try to replace creating dependency by hand by one of Dependency Injection framework e.g. Niject?

File: HotelApi/packages.config [hunk 6]
```diff
@@ -0,0 +1,52 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="Microsoft.AspNet.WebApi.Client" version="5.2.7" targetFramework="net472" />
+  <package id="Microsoft.AspNet.WebApi.Core" version="5.2.7" targetFramework="net472" />
+  <package id="Microsoft.AspNet.WebApi.Owin" version="5.2.7" targetFramework="net472" />
+  <package id="Microsoft.AspNet.WebApi.OwinSelfHost" version="5.2.7" targetFramework="net472" />
+  <package id="Microsoft.AspNetCore.Authentication.Abstractions" version="2.2.0" targetFramework="net472" />
```
Comment (jacekmlynek) [inline HotelApi/packages.config]: Just to note

For me it looks like you have used default project and didn't optimize your dependecy in the project. A lot of them currently are not used. 
Usually we try to keep only these which are currently under the use.

**Review by jacekmlynek — CHANGES_REQUESTED** [review]
Nice and clean solution. Few comments added

File: HotelApi/Startup.cs (hunk 1 above)
Comment (azgas) [reply HotelApi/Startup.cs]: applied

File: HotelApi/Controllers/ReservationController.cs (hunk 2 above)
Comment (azgas) [reply HotelApi/Controllers/ReservationController.cs]: applied

File: HotelApi/Controllers/ReservationController.cs (hunk 3 above)
Comment (azgas) [reply HotelApi/Controllers/ReservationController.cs]: applied

File: HotelApi/Program.cs (hunk 4 above)
Comment (azgas) [reply HotelApi/Program.cs]: applied

File: HotelApi/Controllers/ReservationController.cs (hunk 5 above)
Comment (azgas) [reply HotelApi/Controllers/ReservationController.cs]: applied (after struggling for a couple of hours with nuget packages)

File: HotelApi/packages.config (hunk 6 above)
Comment (azgas) [reply HotelApi/packages.config]: applied

File: HotelApi/packages.config [hunk 7]
```diff
@@ -3,50 +3,13 @@
...
-  <package id="Microsoft.Extensions.DependencyModel" version="2.1.0" targetFramework="net472" />
-  <package id="Microsoft.Extensions.FileProviders.Abstractions" version="2.2.0" targetFramework="net472" />
-  <package id="Microsoft.Extensions.Hosting.Abstractions" version="2.2.0" targetFramework="net472" />
-  <package id="Microsoft.Extensions.Logging.Abstractions" version="2.2.0" targetFramework="net472" />
-  <package id="Microsoft.Extensions.ObjectPool" version="2.2.0" targetFramework="net472" />
-  <package id="Microsoft.Extensions.Options" version="2.2.0" targetFramework="net472" />
-  <package id="Microsoft.Extensions.Primitives" version="2.2.0" targetFramework="net472" />
-  <package id="Microsoft.Net.Http.Headers" version="2.2.0" targetFramework="net472" />
-  <package id="Microsoft.Owin" version="2.0.2" targetFramework="net472" />
-  <package id="Microsoft.Owin.Host.HttpListener" version="2.0.2" targetFramework="net472" />
```
Comment (jacekmlynek) [inline HotelApi/packages.config]: Must have:

This have to stay. It is runtime dependecy and as I can see now your Api will not work without it. I am not sure how you have remove packages, but when you create a clean project you have number of option like `dotnet new webapi --auth none` [More about](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-new?tabs=netcore22). In terms how to clean not needed dependencies ,  when the project already exist, we usually using Resharper, but this tool will also remove runtime depenecies since it is no way to know about them beside human documentation.

In short you have to add this nuget back to the project

File: HotelApi/Controllers/ReservationController.cs [hunk 8]
```diff
@@ -1,40 +1,29 @@
...
-        [HttpGet]
         public JsonResult<List<int>> GetAvailableHotelsIds()
         {
             return Json(_manager.PresentAvailableHotels());
         }
 
-        [HttpPost]
-        public JsonResult<ReservationResult> MakeReservation(int hotelId, double price, int creditCardNumber,
-            string email, DateTime date)
+        public JsonResult<ReservationResult> PostReservation([FromBody] ReservationRequest request)
```
Comment (jacekmlynek) [inline HotelApi/Controllers/ReservationController.cs]: To consider:

Is there any particular reason for FormBody param here? I have checked and it looks like it working without it .

File: HotelApi/Controllers/ReservationRequest.cs [hunk 9]
```diff
@@ -0,0 +1,24 @@
+﻿using System;
+using System.Runtime.Serialization;
+
+namespace HotelApi.Controllers
+{
+    [DataContract]
+    public class ReservationRequest
+    {
+        [DataMember(Name = "hotel_id")]
```
Comment (jacekmlynek) [inline HotelApi/Controllers/ReservationRequest.cs]: Nice to have.

Using DataContract attributes you can specify if field is mandatory or not using IsRequired

**Review by jacekmlynek — CHANGES_REQUESTED** [review]
Looks like a move in the good direction few small comments added.

File: HotelApi/packages.config (hunk 7 above)
Comment (azgas) [reply HotelApi/packages.config]: I used resharper to remove unused dependencies and then somehow I ended up with no packages.config at all - I tried to add the necessary packages again, but this is kind of "magic" to me what exactly each one does. Hopefully now it's better

File: HotelApi/Controllers/ReservationController.cs (hunk 8 above)
Comment (azgas) [reply HotelApi/Controllers/ReservationController.cs]: you're right, I removed it

File: HotelApi/Controllers/ReservationRequest.cs (hunk 9 above)
Comment (azgas) [reply HotelApi/Controllers/ReservationRequest.cs]: applied

**Review by jacekmlynek — APPROVED** [review]
Congratulations your first Api  has been done !!! :)
